### ISSUE TITLE
Exploration of a11y testing options

### ISSUE THREAD

**Issue by ioanacontu** [cds-snc/design-gc-conception#103]
**TODO**
- [x] Research/document shadow DOM and web accessibility, and how we will work with it
- [x] Document tools available for manual a11y testing of components
- [x] Document ways to automate a11y testing working with Stencil
- [x] Document ways to automate a11y testing on alpha document site

**Outcomes**
- Have reference documentation to understand how to work with shadow DOM
- Have a resource of available tools to help ensure a11y is baked into the development process

**Deliverables**
- Document about shadow DOM and web a11y
- List of available tools to manually test a11y
- Document of potential automated a11y solutions for Stencil
- [Web A11y and the Tools to Help Us](https://docs.google.com/presentation/d/1NU0zPZVq0wacLoqj7nwwmns0NMfCn35hSOgBKVOYNlk/edit?usp=sharing)

**Estimated Effort**
- [ ] XS: 1 to 2 hours
- [ ] S: half a day
- [ ] M: 1-2 day
- [x] L: 1 week
- [ ] XL: Move to an epic

**Comment by ethanWallace** [comment]
Shadow DOM and accessibility

The general consensus online is that interactive UI elements can be placed in the shadow DOM and it is the responsibility of the developer to do the additional steps to make it accessible. Wasn't able to find any documentation to go against placing elements like a button in the shadow DOM.

I also ran a test on a normal HTML button, the poc button (gc-ds-button) and another web component button (ionic framework) using two screen readers (VoiceOver and NVDA) to see how all the elements were handled. The screen readers recognized every button the same, normal HTML element and web component.

Conclusion: We shouldn't worry too much about the elements that are placed in the shadow DOM, we just have to make sure the web components are developed with accessibility in mind and follow expected interaction patterns.

**Comment by ethanWallace** [comment]
Setup basic automated tests on the current Stencil GitHub repo. Tests make use of axe-core/puppeteer to run a basic accessibility test on the gc-ds-button component.

[Pull request with passed test](https://github.com/JoshBeveridge/gc-ds-button/pull/1)
[Pull request with failed test](https://github.com/JoshBeveridge/gc-ds-button/pull/2)

**Comment by ethanWallace** [comment]
The alpha documentation site will now run an accessibility test on all pages when a page is modified or published using pa11y. the results of the test are available [https://cds-snc.github.io/alpha-design-system-documentation/reports/](https://cds-snc.github.io/alpha-design-system-documentation/reports/)

**Comment by duboisp** [comment]
Just a side note, automated accessibility reporting will always be partial. Accessibility testing will always require manual assessment.

For example, after a quick review I noticed an accessibility issue in this file: https://github.com/cds-snc/alpha-design-system-documentation/blob/main/src/_includes/partials/nav.njk
where the `<section>` element is useless because a `<nav>` is also a section. And where the name of the section should be instead moved as a heading for the navigation element.
Also, some section lacks heading, see the three (3) HTML validation warning: https://validator.nu/?doc=https%3A%2F%2Fcds-snc.github.io%2Falpha-design-system-documentation%2Fen%2F

I opened a PR to fix those issue: https://github.com/cds-snc/alpha-design-system-documentation/pull/1
